### PR TITLE
🎉  Introduce Google Analytics Data API source

### PR DESCRIPTION
**Pull request by mohamagdy**
## What
Introduce  [Google Analytics Data API source](https://developers.google.com/analytics/devguides/reporting/data/v1)

## 🚨 User Impact 🚨
No user impact expected as this is a new source connector.

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

## Tests

<details><summary><strong>Unit</strong></summary>

```shell
cachedir: .pytest_cache
rootdir: /Users/mohamedmagdy/projects/airbyte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ...
 airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_source.py::test_check_connection ✓                                                                                                            33% ███▍
 airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_source.py::test_discover ✓                                                                                                                    67% ██████▋
 airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/unit_test.py::test_example_method ✓                                                                                                               100% ██████████
============================================================================================================== warnings summary ==============================================================================================================
../../../../../../.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/sources/utils/transform.py:12
  /Users/mohamedmagdy/.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/sources/utils/transform.py:12: DeprecationWarning: Call to deprecated class AirbyteLogger. (Use logging.getLogger('airbyte') instead) -- Deprecated since version 0.1.47.
    logger = AirbyteLogger()

../../../../../../.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/rate_limiting.py:19
  /Users/mohamedmagdy/.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/rate_limiting.py:19: DeprecationWarning: Call to deprecated class AirbyteLogger. (Use logging.getLogger('airbyte') instead) -- Deprecated since version 0.1.47.
    logger = AirbyteLogger()

../../../../../../.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/utils/event_timing.py:13
  /Users/mohamedmagdy/.pyenv/versions/3.9.7/envs/source-google-analytics-data-api/lib/python3.9/site-packages/airbyte_cdk/utils/event_timing.py:13: DeprecationWarning: Call to deprecated class AirbyteLogger. (Use logging.getLogger('airbyte') instead) -- Deprecated since version 0.1.47.
    logger = AirbyteLogger()

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (1.33s):
       3 passed
```

</details>

<details><summary><strong>Integration</strong></summary>

```shell
cachedir: .pytest_cache
rootdir: /Users/mohamedmagdy/projects/airbyte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ...

Results (0.04s):
```

</details>

<details><summary><strong>Acceptance</strong></summary>

```shell
Test session starts (platform: darwin, Python 3.9.7, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/mohamedmagdy/projects/airbyte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ...
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_config_match_spec[inputs0] ✓                                                                    5% ▌
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                                                                       9% ▉
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_docker_env[inputs0] ✓                                                                          14% █▍
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oneof_usage[inputs0] ✓                                                                         18% █▊
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                                                                            23% ██▍
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                                                                            27% ██▊
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                                                                          32% ███▎
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓                                                          36% ███▋
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_defined_refs_exist_in_json_spec_file[inputs0] ✓                                                41% ████▏
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                                                               45% ████▋
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                                                                         50% █████
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                                                                         55% █████▌
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                                                                       59% █████▉
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓                                                64% ██████▍
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs0] ✓                                                   68% ██████▊
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-allOf] ✓                                          73% ███████▍
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-not] ✓                                            77% ███████▊
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_primary_keys_exist_in_schema[inputs0] ✓                                                   82% ████████▎
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                                                                           86% ████████▋
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                                                     91% █████████▏
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                                                  95% █████████▋
 ../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] s                                   100% ██████████
{"type": "LOG", "log": {"level": "INFO", "message": "/Users/mohamedmagdy/projects/airbyte/airbyte/airbyte-integrations/connectors/source-google-analytics-data-api - SAT run - 27e6ce2ca82ea7088b4df2e1531eeb439fa2885c - PASSED"}}

========================================================================================================== short test summary info ===========================================================================================================
SKIPPED [1] ../../../../../canary/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py:21: `future_state_path` not specified, skipping

Results (58.06s):
      21 passed
       1 skipped
```

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            order_bys=[
+                OrderBy(
+                    dimension=OrderBy.DimensionOrderBy(
+                        dimension_name=DEFAULT_CURSOR_FIELD,
+                        order_type=OrderBy.DimensionOrderBy.OrderType.ALPHANUMERIC
+                    )
+                )
+            ]
+        )
+        return client.run_report(request)
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py:180]: can we create client.py a client(GoogleAnalyticsData) class and add this there

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        response = self._run_report(config)
+
+        dimensions = list(map(lambda h: h.name, response.dimension_headers))
+        metrics = list(map(lambda h: h.name, response.metric_headers))
+
+        last_cursor_value = state.get(report_name, {}).get(DEFAULT_CURSOR_FIELD, "")
+
+        for row in response.rows:
+            data = dict(zip(dimensions, list(map(lambda v: v.value, row.dimension_values))))
+            data.update(dict(zip(metrics, list(map(lambda v: float(v.value), row.metric_values)))))
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py:135]: we can move these data processes to that client

**Review by harshithmullapudi — CHANGES_REQUESTED** [review]
Also
1. can you run `./gradlew format`
2. share the screenshot for integration tests

**Comment by harshithmullapudi** [comment]
@mohamagdy any update here?

**Comment by mohamagdy** [comment]
> return client.run_report(request)

**Comment by mohamagdy** [comment]
> @mohamagdy any update here?

Yes

> 1. can you run `./gradlew format`

I got this error. Tried to figure out what is the issue but didn't find any as far as I can see. The imports look good in that file. 
```shell
> Task :spotlessJava FAILED
Step 'removeUnusedImports' found problem in 'airbyte-analytics/src/main/java/io/airbyte/analytics/SegmentTrackingClient.java':
null
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$constructRemoveUnusedFunction$4(GoogleJavaFormatStep.java:210)
        at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$createRemoveUnusedImportsOnly$2(GoogleJavaFormatStep.java:187)
```

> 2. share the screenshot for integration tests

```shell
➜  source-google-analytics-data-api git:(source-google-analytics-data-api) ✗ python -m pytest integration_tests
Test session starts (platform: darwin, Python 3.9.7, pytest 6.2.5, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/mohamedmagdy/projects/airbyte/airbyte, configfile: pytest.ini
plugins: sugar-0.9.4, mock-3.6.1, timeout-1.4.2
collecting ...

Results (0.03s):
```

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py (hunk 2 above)
Comment (mohamagdy) [reply airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py]: Moved here https://github.com/airbytehq/airbyte/pull/12701/commits/95b2855db98c01c174bf527a3e83fbafb5285f6a.

**Comment by misteryeo** [comment]
Hi folks, would it be accurate to say that this is a source for Google Analytics 4 (which leverages the Data API?)

**Comment by mohamagdy** [comment]
Hey @misteryeo, I would say the Google Anayltics v4 and the Google Analytics Data API are 2 separate things but the Google Analytics Data API gives access indirectly to Google Analytics v4 data via a "reporting" API based on the definition here https://developers.google.com/analytics/devguides/reporting/data/v1.

**Review by harshithmullapudi — APPROVED** [review]

**Comment by harshithmullapudi** [comment]
/publish connector=connectors/source-google-analytics-data-api auto-bump-version=false

**Comment by harshithmullapudi** [comment]
/publish connector=connectors/source-google-analytics-data-api auto-bump-version=false

**Comment by harshithmullapudi** [comment]
/publish connector=connectors/source-google-analytics-data-api auto-bump-version=false

**Comment by girarda** [comment]
This issue was closed because it was deprioritized. Feel free to reopen if needed.

**Pull request merged**